File: plugins/qr-scanner/ios/QRScannerPlugin.swift
//
//  QRScannerPlugin.swift — lecodes-plugins/qr-scanner
//
//  The "qrScanner" registerView plugin: pairs with the SDK's QRScanner() wrapper, which
//  rides the generic NativeView channel — no per-plugin bridge methods.
//

import UIKit
import LeCodesSDK

public enum LeCodesQRScannerPlugin {

    public static let pluginId = "qr-scanner"

    public static func register(in engine: LeCodesEngine) {
        engine.registerView("qrScanner") { _, channel in
            QRScannerViewInstance(channel)
        }
    }
}

/// The "qrScanner" registerView factory product: wraps QRScannerView, forwarding decoded
/// codes as `scan { data }` events (data: null when the code leaves the frame).
final class QRScannerViewInstance: NativeViewInstance {

    let view: UIView

    init(_ channel: NativeViewChannel) {
        view = QRScannerView(onData: { data in
            channel.emit("scan", ["data": (data as Any?) ?? NSNull()])
        })
    }
}

File: plugins/qr-scanner/android/src/main/java/io/letary/lecodes/plugins/qrscanner/QRScannerPlugin.kt
//
//  QRScannerPlugin.kt — lecodes-plugins/qr-scanner
//
//  The "qrScanner" registerView plugin. Event surface matches iOS and the SDK QRScanner
//  wrapper exactly: only "scan" { data } is emitted (data null when the code leaves the
//  frame) — there is no "open" event in the contract.
//

package io.letary.lecodes.plugins.qrscanner

import android.content.Context
import io.letary.lecodes.LecodesEngine
import io.letary.lecodes.utils.NativeViews
import org.json.JSONObject

object QRScannerPlugin {

    const val PLUGIN_ID = "qr-scanner"

    fun register(engine: LecodesEngine, context: Context) {
        val appContext = context.applicationContext
        engine.registerView("qrScanner") { _, channel ->
            val view = QRScannerView(appContext,
                onOpen = {},
                onData = { data -> channel.emit("scan", JSONObject().put("data", data ?: JSONObject.NULL).toString()) })
            NativeViews.Instance(view)
        }
    }
}
